pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -104,6 +104,21 @@ type BMCDetails struct {
 	CredentialsName string `json:"credentialsName"`
 }
 
+// BIOSConfig contains the configuration that you want to configure BIOS settings in Bare metal server
+type BIOSConfig struct {
+
+	// Select the boot mode of the system. Allowed values are Uefi, LegacyBios
+	// +kubebuilder:validation:Enum=Uefi,LegacyBios
+	BootMode string `json:"bootMode,omitempty"`
+
+	// Hyper-threading technology allows a single physical processor core to appear as several
+	// logical processors. This supports following options: true, false.
+	HyperthreadingEnabled bool `json:"hyperthreadingEnabled,omitempty"`
```

Comment:
Do all processor types support this?

I'm thinking about how many different axes we have for deciding if a BIOS setting is relevant to a given host, and whether we need some way to separate the ones that aren't applicable in some cases. Do we want to put them in their own struct (with a name like DellIntelBIOS or DellR640BIOS or something) *or* if we want to just keep the option list very short, abstract things as you've done here, and have the implementation ignore anything that isn't relevant for its platform. I could see either approach working.

Reply:
> If we want to put some optional BIOS setting that depend on a given host (given hardware driver), I think we can declare `extraArgs` in CRD with `interface` type as I have done in previous PR:
> 
> ```go
> extraArgs interface{} `json:"optionalSettings,omitempty"`
> ```
> 
> The YAML file may look like:
> 
> ```yaml
>   bios:
>     bootMode: uefi
>     extraArgs:
>       setting1: "value1"
>       setting2: "value2"
> ```
> 
> But in this case, the validator for these settings must be defined in indicated hardware driver (iRMC, iLO, ...), not in metal3's API.
> 
> How to you think about it?

No, we can't do that. It's bad API design to have informal fields like that, but beyond the design issue I know that in the near future kubernetes is going to require more formal validation for CRDs to ensure that the structure of the data passed in matches what is expected and does not include any "extra" fields. This requirement was presented as addressing a security issue that comes up when someone stores extra fields in a resource today and then in the future that resource definition changes to include the fields, because the existing data would never have been validated.

So, we must find a way to do this by declaring everything explicitly.